### PR TITLE
New setting to prevent automatically importing dangling indices

### PR DESCRIPTION
**Pull request by pugnascotia**
Part of #48366.

Introduce a new static setting, `gateway.auto_import_dangling_indices`, which prevents dangling indices from being automatically imported.

I've also updated the dangling index docs a little to cover the new setting, and add some scary caveats.

### CODE REVIEW

**Comment by pugnascotia** [comment]
Note: there's more to do to satisfy the issue, e.g. an API to list the danglies, but I plan on doing that separately.

**Comment by pugnascotia** [comment]
@DaveCTurner @original-brownbear ready for a review, if you have time.

**Review by ywelsch — COMMENTED** [review]
Talked to @DaveCTurner who will add some comments as well. No  need to revise until then

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
Thanks @pugnascotia, I left a few suggestions.

**Comment by pugnascotia** [comment]
Thanks for the reviews, I've pushed 71bcbde1b6f1e646952e1d017649fe6d2c161a3a to address them, specifically:

   * Reverted the docs changes
   * Default the new setting to false
   * Only subscribe `DanglingIndicesState` to cluster state updates if the new setting is true
   * Revert changes to `UnsafeBootstrapAndDetachCommandIT` as they're no longer needed
   * Strip confusing comments from `DanglingIndicesStateTests` and be specify the new setting's value when calling `createDanglingIndicesState()`.

**Comment by pugnascotia** [comment]
@ywelsch @DaveCTurner any chance of a re-review?

File: server/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java [hunk 1]
```diff
@@ -55,19 +56,35 @@
...
         this.allocateDangledIndices = allocateDangledIndices;
-        clusterService.addListener(this);
+        this.allocateDanglingIndices = AUTO_IMPORT_DANGLING_INDICES_SETTING.get(clusterService.getSettings());
+
+        if (this.allocateDanglingIndices) {
+            clusterService.addListener(this);
+        }
+    }
+
+    public void setAllocateDanglingIndicesSetting(boolean allocateDanglingIndices) {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java]: this method is superfluous now?

File: server/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java [hunk 2]
```diff
@@ -171,10 +188,15 @@ private IndexMetaData stripAliases(IndexMetaData indexMetaData) {
      * Allocates the provided list of the dangled indices by sending them to the master node
      * for allocation.
      */
-    private void allocateDanglingIndices() {
+    void allocateDanglingIndices() {
+        if (this.allocateDanglingIndices == false) {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java]: this condition is superfluous now?

File: server/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java [hunk 3]
```diff
@@ -55,19 +56,35 @@
...
+        Setting.Property.NodeScope
+    );
+
     private final NodeEnvironment nodeEnv;
     private final MetaStateService metaStateService;
     private final LocalAllocateDangledIndices allocateDangledIndices;
 
     private final Map<Index, IndexMetaData> danglingIndices = ConcurrentCollections.newConcurrentMap();
 
+    private boolean allocateDanglingIndices;
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java]: this field is superfluous now?

File: server/src/test/java/org/elasticsearch/gateway/DanglingIndicesStateTests.java [hunk 4]
```diff
@@ -35,16 +36,21 @@
...
+import static org.mockito.Mockito.when;
 
 public class DanglingIndicesStateTests extends ESTestCase {
 
     private static Settings indexSettings = Settings.builder()
-            .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, 1)
-            .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, 0)
-            .put(IndexMetaData.SETTING_VERSION_CREATED, Version.CURRENT)
-            .build();
+        .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, 1)
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/gateway/DanglingIndicesStateTests.java]: Let's avoid reformatting code unrelated to the PR.

File: server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+    private void createAndPopulateIndex(String name, int shardCount, int replicaCount) throws InterruptedException {
+        logger.info("--> creating test index: {}", name);
+        assertAcked(
+            prepareCreate(
+                name,
+                Settings.builder()
+                    .put("number_of_shards", shardCount)
+                    .put("number_of_replicas", replicaCount)
+                    .put(Store.INDEX_STORE_STATS_REFRESH_INTERVAL_SETTING.getKey(), 0)
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java]: why is this necessary?

File: server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+import static org.hamcrest.Matchers.equalTo;
+
+@ClusterScope(numDataNodes = 0, scope = ESIntegTestCase.Scope.TEST)
+public class DanglingIndicesIT extends ESIntegTestCase {
+    private static final String INDEX_NAME = "test-idx-1";
+
+    private static final int MIN_DOC_COUNT = 500;
+    private static final int SHARD_COUNT = 1;
+    private static final int REPLICA_COUNT = 2;
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java]: why have these constants here? If you're always passing the same shard and replica count to `createAndPopulateIndex`, there is no point in having these. I find there to be a lot of ceremony in this class that is not adding anything.

File: server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+                Settings.builder()
+                    .put("number_of_shards", shardCount)
+                    .put("number_of_replicas", replicaCount)
+                    .put(Store.INDEX_STORE_STATS_REFRESH_INTERVAL_SETTING.getKey(), 0)
+            )
+        );
+        ensureGreen();
+
+        logger.info("--> indexing sample data");
+        final int numDocs = between(MIN_DOC_COUNT, 1000);
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java]: no need to add any docs. Let's keep it simple

File: server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+        final IndexRequestBuilder[] docs = new IndexRequestBuilder[numDocs];
+
+        for (int i = 0; i < numDocs; i++) {
+            docs[i] = client().prepareIndex(name)
+                .setSource("foo-int", randomInt(), "foo-string", randomAlphaOfLength(32), "foo-float", randomFloat());
+        }
+
+        indexRandom(true, docs);
+        flush();
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java]: why flush?

File: server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+        assertThat(client().prepareSearch(name).setSize(0).get().getHits().getTotalHits().value, equalTo((long) numDocs));
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java]: why check that indexing works ok? Not the point of this test, that is tested elsewhere.

File: server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+
+        client().admin().indices().prepareStats(name).execute().actionGet();
+    }
+
+    private void deleteIndex(String indexName) {
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java]: why define an extra method here? just so that one logging line can be reused? Feels a bit over the top.

File: test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java [hunk 11]
```diff
@@ -1467,21 +1469,22 @@ public int size() {
     }
 
     /**
-     * Stops a random data node in the cluster. Returns true if a node was found to stop, false otherwise.
+     * Stops a random data node in the cluster and removes it.
+     * @return the name of the stopped node, if a node was found to stop.
      */
-    public synchronized boolean stopRandomDataNode() throws IOException {
+    public synchronized Optional<String> stopRandomDataNode() throws IOException {
```
Comment (ywelsch) [inline test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java]: where is this change used?

File: test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java [hunk 12]
```diff
@@ -1902,6 +1905,12 @@ public String startNode(Settings settings) {
         return startNodes(settings).get(0);
     }
 
+    public void startNode(String nodeName) {
```
Comment (ywelsch) [inline test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java]: where is this used? Revert

File: server/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java [hunk 13]
```diff
@@ -55,19 +56,35 @@
 
     private static final Logger logger = LogManager.getLogger(DanglingIndicesState.class);
 
+    public static final Setting<Boolean> AUTO_IMPORT_DANGLING_INDICES_SETTING = Setting.boolSetting(
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java]: Let's deprecate this setting right away, so that we can remove it in 8.0 if we want

File: server/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java [hunk 14]
```diff
@@ -55,19 +56,35 @@
...
     public DanglingIndicesState(NodeEnvironment nodeEnv, MetaStateService metaStateService,
                                 LocalAllocateDangledIndices allocateDangledIndices, ClusterService clusterService) {
         this.nodeEnv = nodeEnv;
         this.metaStateService = metaStateService;
         this.allocateDangledIndices = allocateDangledIndices;
-        clusterService.addListener(this);
+        this.allocateDanglingIndices = AUTO_IMPORT_DANGLING_INDICES_SETTING.get(clusterService.getSettings());
+
+        if (this.allocateDanglingIndices) {
+            clusterService.addListener(this);
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java]: with this change, we are now not scanning for new dangling indices anymore and therefore also not warning about the existence of dangling indices on disk. We should keep this in mind when working out a new story for managing dangling indices via an API. Perhaps we should think about still emitting a warning in the logs. Let's add a TODO for this on the meta issue.

**Review by ywelsch — CHANGES_REQUESTED** [review]
Thanks for the update. I've left some comments.

File: server/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java (hunk 1 above)
Comment (pugnascotia) [reply server/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java]: Oops, yes, good catch.

File: server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java (hunk 5 above)
Comment (pugnascotia) [reply server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java]: Ahem...that's copy-and-paste. Can you tell it's the first time I've written an ES IT? 😅

**Comment by pugnascotia** [comment]
@ywelsch thank you for the review - seems there was quite a bit of cruft to clear up, sorry about that. Notes on the latest changes:

   * The new setting is now deprecated
   * The state class logs a warning if the setting is false, since dangling indices won't be detected or imported
   * Change the disabled case unit test to verify that no cluster state listener is registered, since the `danglingState.allocateDanglingIndices()` method no longer checks the setting value
   * Tidied up the IT

File: server/src/test/java/org/elasticsearch/gateway/DanglingIndicesStateTests.java [hunk 15]
```diff
@@ -46,6 +51,12 @@
             .put(IndexMetaData.SETTING_VERSION_CREATED, Version.CURRENT)
             .build();
 
+    // The setting AUTO_IMPORT_DANGLING_INDICES_SETTING is deprecated, so we must disable
+    // warning checks or all the tests will fail.
+    protected boolean enableWarningsCheck() {
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/gateway/DanglingIndicesStateTests.java]: @Override

File: server/src/test/java/org/elasticsearch/gateway/DanglingIndicesStateTests.java [hunk 16]
```diff
@@ -181,7 +191,66 @@ public void testDanglingIndicesStripAliases() throws Exception {
...
+        try (NodeEnvironment env = newNodeEnvironment()) {
+            MetaStateService metaStateService = new MetaStateService(env, xContentRegistry());
+            LocalAllocateDangledIndices localAllocateDangledIndices = mock(LocalAllocateDangledIndices.class);
+
+            final ClusterService mockClusterService = buildMockClusterService(false);
+
+            createDanglingIndicesState(env, metaStateService, localAllocateDangledIndices, mockClusterService);
+
+            // Check that no listener was registered, because auto-import is disabled
+            verify(mockClusterService, never()).addListener(any());
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/gateway/DanglingIndicesStateTests.java]: I'm not a big fan of these whitebox tests, given how fragile they are. There is also no test here that checks that `addListener` is ever called when auto-import is enabled? This means that if the dangling indices implementation changes to use another method on clusterservice to register a callback, this test will be pointless.

File: server/src/test/java/org/elasticsearch/gateway/DanglingIndicesStateTests.java [hunk 17]
```diff
@@ -181,7 +191,66 @@ public void testDanglingIndicesStripAliases() throws Exception {
...
+        return createDanglingIndicesState(env, metaStateService, null, buildMockClusterService(true));
+    }
+
+    private DanglingIndicesState createDanglingIndicesState(
+        NodeEnvironment env,
+        MetaStateService metaStateService,
+        LocalAllocateDangledIndices indexAllocator,
+        ClusterService clusterService
+    ) {
+        return new DanglingIndicesState(env, metaStateService, indexAllocator, clusterService);
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/gateway/DanglingIndicesStateTests.java]: what's the point of this method? Let's avoid these unnecessary abstractions.

File: server/src/test/java/org/elasticsearch/gateway/DanglingIndicesStateTests.java [hunk 18]
```diff
@@ -181,7 +191,66 @@ public void testDanglingIndicesStripAliases() throws Exception {
         }
     }
 
+    public void testDanglingIndicesAreNotAllocatedWhenDisabled() throws Exception {
+        try (NodeEnvironment env = newNodeEnvironment()) {
+            MetaStateService metaStateService = new MetaStateService(env, xContentRegistry());
+            LocalAllocateDangledIndices localAllocateDangledIndices = mock(LocalAllocateDangledIndices.class);
+
+            final ClusterService mockClusterService = buildMockClusterService(false);
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/gateway/DanglingIndicesStateTests.java]: I would find the test easier to understand if we were to explicitly pass the setting `Settings.builder().put(AUTO_IMPORT_DANGLING_INDICES_SETTING.getKey(), false).build()` here instead of the magic boolean.

File: server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+     * the cluster, so long as the recovery setting is enabled.
+     */
+    public void testDanglingIndicesAreRecoveredWhenSettingIsEnabled() throws Exception {
+        logger.info("--> starting cluster");
+        final Settings settings = buildSettings(true);
+        internalCluster().startNodes(3, settings);
+
+        // Create an index and distribute it across the 3 nodes
+        createAndPopulateIndex(INDEX_NAME, 1, 2);
+        ensureGreen();
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java]: `createAndPopulateIndex` already has a call to `ensureGreen`.

File: server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+                return super.onNodeStopped(nodeName);
+            }
+        });
+
+        ensureGreen();
+
+        assertFalse("Did not expect dangling index " + INDEX_NAME + " to be recovered", indexExists(INDEX_NAME));
+    }
+
+    private void createAndPopulateIndex(String name, int shardCount, int replicaCount) {
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java]: I would prefer not to have this method, just do the actual calls 

Note that there is also the `createIndex(String name, Settings indexSettings)` on ESIntegTestCase which allows to write this succinctly.

File: server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+        logger.info("--> creating test index: {}", name);
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java]: the logging here is not necessary. Elasticsearch already logs index creations at info level.

File: server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+     */
+    public void testDanglingIndicesAreNotRecoveredWhenSettingIsDisabled() throws Exception {
+        logger.info("--> starting cluster");
+        internalCluster().startNodes(3, buildSettings(false));
+
+        // Create an index and distribute it across the 3 nodes
+        createAndPopulateIndex(INDEX_NAME, 1, 2);
+
+        // Create another index so that once we drop the first index, we
+        // can still assert that the cluster is green.
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java]: I don't understand this.`ensureGreen` does not require for an index to be present?

File: server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java [hunk 23]
```diff
@@ -0,0 +1,114 @@
...
+        createAndPopulateIndex(INDEX_NAME, 1, 2);
+        ensureGreen();
+
+        // This is so that when then node comes back up, we have a dangling index that can be recovered.
+        logger.info("--> restarted a random node and deleting the index while it's down");
+        internalCluster().restartRandomDataNode(new InternalTestCluster.RestartCallback() {
+
+            @Override
+            public Settings onNodeStopped(String nodeName) throws Exception {
+                logger.info("--> deleting test index: {}", INDEX_NAME);
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java]: deletions are already logged by ES at info level, no need to log here.

File: server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java [hunk 24]
```diff
@@ -0,0 +1,114 @@
...
+                assertAcked(client().admin().indices().prepareDelete(INDEX_NAME));
+                return super.onNodeStopped(nodeName);
+            }
+        });
+
+        ensureGreen();
+
+        assertTrue("Expected dangling index " + INDEX_NAME + " to be recovered", indexExists(INDEX_NAME));
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java]: There is no guarantee that the dangling index has been recovered at this point, as that has an asynchronous notification / import process. This line needs to be put into an `assertBusy`

File: test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java [hunk 25]
```diff
@@ -1473,7 +1473,8 @@ public int size() {
     }
 
     /**
-     * Stops a random data node in the cluster. Returns true if a node was found to stop, false otherwise.
+     * Stops a random data node in the cluster and removes it.
+     * @return the name of the stopped node, if a node was found to stop.
```
Comment (ywelsch) [inline test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java]: revert this.

File: server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java [hunk 26]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    /**
+     * Check that when dangling indices are discovered, then they are not recovered into
+     * the cluster when the recovery setting is disabled.
+     */
+    public void testDanglingIndicesAreNotRecoveredWhenSettingIsDisabled() throws Exception {
+        logger.info("--> starting cluster");
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java]: what does this logging message add? It's the first line in the test, which is already logged.

File: server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java [hunk 27]
```diff
@@ -0,0 +1,114 @@
...
+
+        // Create another index so that once we drop the first index, we
+        // can still assert that the cluster is green.
+        createAndPopulateIndex(INDEX_NAME + "-other", 1, 2);
+
+        ensureGreen();
+
+        // This is so that when then node comes back up, we have a dangling index that could
+        // be recovered, but shouldn't be.
+        logger.info("--> restarted a random node and deleting the index while it's down");
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java]: The restart is already logged at info level. Maybe just change the comment to say `// Restart node, deleting the index in its absence, so that there is a dangling index to recover`

**Review by ywelsch — CHANGES_REQUESTED** [review]
Thanks for the update @pugnascotia. I've left more comments.

**Comment by pugnascotia** [comment]
Thanks again @ywelsch for bearing with me.

   * I've replaced the white box tests with a getter for the new setting. I had avoided that before because the only code calling the getter would be test code, but it'll be less fragile.
   * I've simplified the unit tests and ITs.

File: server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java [hunk 28]
```diff
@@ -0,0 +1,95 @@
...
+
+    /**
+     * Check that when dangling indices are discovered, then they are recovered into
+     * the cluster, so long as the recovery setting is enabled.
+     */
+    public void testDanglingIndicesAreRecoveredWhenSettingIsEnabled() throws Exception {
+        final Settings settings = buildSettings(true);
+        internalCluster().startNodes(3, settings);
+
+        createIndex(INDEX_NAME, Settings.builder().put("number_of_shards", 1).put("number_of_replicas", 2).build());
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java]: does the number of (primary) shards matter? I understand that you want 2 replicas so that every node has a shard copy.

File: server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java [hunk 29]
```diff
@@ -0,0 +1,95 @@
...
+                assertAcked(client().admin().indices().prepareDelete(INDEX_NAME));
+                return super.onNodeStopped(nodeName);
+            }
+        });
+
+        // Since index recovery is async, we can't prove index recovery will never occur, just that it doesn't occur within some reasonable
+        // amount of time
+        assertFalse(
+            "Did not expect dangling index " + INDEX_NAME + " to be recovered",
+            waitUntil(() -> indexExists(INDEX_NAME), 5, TimeUnit.SECONDS)
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java]: Let's reduce this to 1 second

File: server/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java [hunk 30]
```diff
@@ -67,7 +76,18 @@ public DanglingIndicesState(NodeEnvironment nodeEnv, MetaStateService metaStateS
...
+        this.isAutoImportDanglingIndicesEnabled = AUTO_IMPORT_DANGLING_INDICES_SETTING.get(clusterService.getSettings());
+
+        if (this.isAutoImportDanglingIndicesEnabled) {
+            clusterService.addListener(this);
+        } else {
+            logger.warn(AUTO_IMPORT_DANGLING_INDICES_SETTING.getKey() + " is disabled, dangling indices will not be detected or imported");
+        }
+    }
+
+    public boolean isAutoImportDanglingIndicesEnabled() {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java]: does not need to be public, only accessed by tests

**Review by ywelsch — APPROVED** [review]
LGTM
